libcloud/storage/base.py

```diff
@@ -622,200 +624,62 @@ def _upload_object(self, object_name, content_type, upload_func,
-                    response.connection.connection.send(b('%X\r\n' %
-                                                          (len(chunk))))
-                    response.connection.connection.send(b(chunk))
-                    response.connection.connection.send(b('\r\n'))
+        if upload_func:
+            upload_func(**upload_func_kwargs)
+
+        return {'response': response,
+                'bytes_transferred': stream_length,
+                'data_hash': stream_hash}
+
+    def _hash_buffered_stream(self, stream, hasher, blocksize=65536):
+        total_len = 0
+        if hasattr(stream, '__next__'):
+            data = libcloud.utils.files.exhaust_iterator(iterator=stream)
+            hasher.update(b(data))
+            total_len = len(data)
+            return (hasher.hexdigest(), total_len)
+        if not hasattr(stream, '__exit__'):
+            for s in stream:
+                if isinstance(s, str):
+                    hasher.update(s)
                 else:
-                    response.connection.connection.send(b(chunk))
-            except Exception:
-                # TODO: let this exception propagate
-                # Timeout, etc.
-                return False, None, bytes_transferred
-
-            bytes_transferred += len(chunk)
-            if calculate_hash:
-                data_hash.update(b(chunk))
-
-            try:
-                chunk = next(generator)
-            except StopIteration:
-                chunk = ''
-
-        if chunked:
-            response.connection.connection.send(b('0\r\n\r\n'))
-
-        if calculate_hash:
-            data_hash = data_hash.hexdigest()
-
-        return True, data_hash, bytes_transferred
-
-    def _upload_file(self, response, file_path, chunked=False,
-                     calculate_hash=True):
-        """
-        Upload a file to the server.
-
-        :type response: :class:`RawResponse`
-        :param response: RawResponse object.
-
-        :type file_path: ``str``
-        :param file_path: Path to a local file.
-
-        :type iterator: :class:`object`
-        :param response: An object which implements an iterator interface (File
-                         object, etc.)
-
-        :rtype: ``tuple``
-        :return: First item is a boolean indicator of success, second
-                 one is the uploaded data MD5 hash and the third one
-                 is the number of transferred bytes.
-        """
-        with open(file_path, 'rb') as file_handle:
-            success, data_hash, bytes_transferred = (
-                self._stream_data(
-                    response=response,
-                    iterator=iter(file_handle),
-                    chunked=chunked,
-                    calculate_hash=calculate_hash))
-
-        return success, data_hash, bytes_transferred
+                    hasher.update(s)
+                total_len = total_len + len(s)
+            return (hasher.hexdigest(), total_len)
+        with stream:
```

Comment:
This block can read:

```python
with stream:  # why though?
    for buf in iter(partial(stream.read, blocksize), ''):
          hasher.update(buf)
          total_len = total_len + len(buf)
```

This makes it easier to fold into the case where you use `for` already.